content/blog/2022-05-19-hcl-mlir/index.md

```diff
@@ -0,0 +1,514 @@
+                   affine.store %sum, %C[%i, %j] : memref<1024x1024xf32>
+               } { loop_name = "k", reduction = 1 : i32}
+           } { loop_name = "j" }
+       } { loop_name = "i", stage_name = "s" }
+       hcl.reorder(%s, %lk, %lj)
+       %buf = hcl.buffer_at(%s, %C: memref<1024x1024xf32>, %li) -> memref<1024xf32>
+       hcl.pipeline(%s, %lj, 1)
+       return
+   }
+}
+```
+
+With the same pass to implement loop reordering and buffer generation, the transformed MLIR assembly code has the reduction loop at axis 1, inner loop pipelined, and partial results written to the row buffer `%0`, as shown in the following code snippet.
+
+```cpp
+module {
+  func @kernel(%arg0: memref<1024x512xf32>, %arg1: memref<512x1024xf32>, %arg2: memref<1024x1024xf32>) {
+    affine.for %arg3 = 0 to 1024 {
+      %0 = memref.alloc() : memref<1024xf32>
+      %cst = arith.constant 0.000000e+00 : f32
+      affine.for %arg4 = 0 to 1024 {
+        affine.store %cst, %0[%arg4] : memref<1024xf32>
+      } {loop_name = "j_init", pipeline_ii = 1 : i32}
+      affine.for %arg4 = 0 to 512 {
+        affine.for %arg5 = 0 to 1024 {
+          %1 = affine.load %arg0[%arg3, %arg4] : memref<1024x512xf32>
+          %2 = affine.load %arg1[%arg4, %arg5] : memref<512x1024xf32>
+          %3 = affine.load %0[%arg5] : memref<1024xf32>
+          %4 = arith.mulf %1, %2 : f32
+          %5 = arith.addf %4, %3 : f32
+          affine.store %5, %0[%arg5] : memref<1024xf32>
+        } {loop_name = "j", pipeline_ii = 1 : i32}
+      } {loop_name = "k", reduction = 1 : i32}
+      affine.for %arg4 = 0 to 1024 {
+        %1 = affine.load %0[%arg4] : memref<1024xf32>
+        affine.store %1, %arg2[%arg3, %arg4] : memref<1024x1024xf32>
+      } {loop_name = "j_back", pipeline_ii = 1 : i32}
+    } {loop_name = "i", stage_name = "s"}
+    return
+  }
+}
+```
+
+## Roofline Model
+The Roofline Model[^5] is an intuitive visual performance model used to provide performance estimates of a given compute kernel running on a computing platform. The hardware limitation is illustrated as the "ceiling" in the roofline model, as shown in Fig. 5. The diagonal roof describes the peak bandwidth, and the horizontal line describes the peak performance of the specific architecture. Performance saturates at the ridge point where the diagonal and horizontal roof meet. 
+
+<center>
+<img src="roofline_intro.png" alt="alt_text" title="image_tooltip" style="zoom:20%;">
+</center>
+<center>
+Fig. 5: Roofline Model
+</center>
+
+<!-- What we do to access the memory access / ops -->
+We add a profiling tool to evaluate the operational intensity by analyzing the MLIR code to count the number of memory access and arithmetic operations. Then a roofline model is generated to guide the optimization process. Specifically, we visit each operation and record the current loop nest. When an arithmetic operation or load/store operation is visited, the number of trip counts in the current loop is added. 
```

Comment:
I'm a tiny bit confused about the high-level idea here: it's called "profiling," but profilers are usually dynamic (i.e., they run a program and measure what it does at run time). This sounds like a static analysis—is that right? If so, is there any worry about precision? Or are all loops statically bounded anyway, so the number of accesses, etc., are always perfectly statically analyzable? If that's the case, it seems like this is worth a mention.

Reply:
Yes, there are two parts in the roofline model:
1. The number of memory accesses and the number of arithmetic operations (FLOPs) can be statically calculated from the source MLIR code. Right now we only support constant loop bounds, but if complex control flow is involved, we probably need to do some instrumentation to dynamically count the numbers.
2. The performance in the y-axis needs to run the HLS and obtain the latency, thus cannot be statically determined.

Reply:
Got it. To be clear, however, I would still call that HLS estimate "static"… in that it doesn't involve _actually running_ the program, which would be necessary if there were input-dependent loop trip counts, for example.